tasm-lib/src/arithmetic/u128/overflowing_add.rs

```diff
@@ -0,0 +1,289 @@
+                // _ rhs_3 rhs_2 lhs_2 sum_0 lhs_3 carry_2 sum_1
+
+                swap 4
+                // _ rhs_3 rhs_2 sum_1 sum_0 lhs_3 carry_2 lhs_2
+
+                add
+                // _ rhs_3 rhs_2 sum_1 sum_0 lhs_3 lhs_2'
+
+                swap 1 swap 4
+                // _ rhs_3 lhs_3 sum_1 sum_0 lhs_2' rhs_2
+
+                add
+                // _ rhs_3 lhs_3 sum_1 sum_0 (lhs_2' + rhs_2)
+
+                split
+                // _ rhs_3 lhs_3 sum_1 sum_0 carry_3 sum_2
+
+                swap 4
+                // _ rhs_3 sum_2 sum_1 sum_0 carry_3 lhs_3
+
+                add
+                // _ rhs_3 sum_2 sum_1 sum_0 lhs_3'
+
+                dup 4
+                // _ rhs_3 sum_2 sum_1 sum_0 lhs_3' rhs_3
+
+                add
+                // _ rhs_3 sum_2 sum_1 sum_0 (lhs_3' + rhs_3)
+
+                split
+                // _ rhs_3 sum_2 sum_1 sum_0 overflow sum_3
+
+                swap 5
+                pop 1
+                // _ sum_3 sum_2 sum_1 sum_0 overflow
+
+                return
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use itertools::Itertools;
+    use num::Zero;
+    use rand::rngs::StdRng;
+    use rand::Rng;
+    use rand::SeedableRng;
+
+    use super::*;
+    use crate::push_encodable;
+    use crate::snippet_bencher::BenchmarkCase;
+    use crate::test_helpers::test_rust_equivalence_given_complete_state;
+    use crate::traits::closure::Closure;
+    use crate::traits::closure::ShadowedClosure;
+    use crate::traits::rust_shadow::RustShadow;
+
+    #[test]
+    fn overflowing_add_u128_test() {
+        ShadowedClosure::new(OverflowingAddU128).test()
+    }
+
+    #[test]
+    fn overflowing_add_u128_unit_test() {
+        let snippet = OverflowingAddU128;
+        snippet.prop_add(1u128 << 67, 1u128 << 67)
+    }
+
+    #[test]
+    fn overflowing_add_u128_overflow_test() {
+        let snippet = OverflowingAddU128;
+
+        for (a, b) in [
+            (1u128 << 127, 1u128 << 127),
+            (u128::MAX, u128::MAX),
+            (u128::MAX, 1),
+            (u128::MAX, 1 << 31),
+            (u128::MAX, 1 << 32),
+            (u128::MAX, 1 << 33),
+            (u128::MAX, 1 << 63),
+            (u128::MAX, 1 << 64),
+            (u128::MAX, 1 << 65),
+            (u128::MAX, 1 << 95),
+            (u128::MAX, 1 << 96),
+            (u128::MAX, 1 << 97),
+            (u128::MAX - 1, 2),
+        ] {
+            snippet.prop_add(a, b);
+            snippet.prop_add(b, a);
+        }
+
+        for i in 0..128 {
+            let a = u128::MAX - ((1u128 << i) - 1);
+            let b = 1u128 << i;
+
+            // sanity check of test input values
+            let (wrapped_add, is_overflow) = a.overflowing_add(b);
+            assert!(is_overflow, "i = {i}. a = {a}, b = {b}");
+            assert!(wrapped_add.is_zero());
+
+            snippet.prop_add(b, a);
+        }
+    }
+
+    impl OverflowingAddU128 {
+        fn prop_add(&self, lhs: u128, rhs: u128) {
```

Comment:
I strongly urge to use a telling function name here. Entire books have been written about the topic of variable identifiers and function names, and for good reason. The fact that this is a test module does not matter; tests have to be read, understood, and maintained just as any other piece of code in this repository, and are not second-class citizens.

A prop is, for example, [“[a]n object placed against or under another, to support it”](https://en.wiktionary.org/wiki/prop), among other meanings, none of which apply. This method adds no prop, contrary to what the name suggests. Since the method panics if a certain property is not met, a better name would be `assert_addition_property`, and there are probably better names still.

Reply:
Commit c33fef02b89d4115768e859b5aeb1a0c96cd007f renames the function to something that describes what it does.

For the sake of pushing back: I think it is unfair to imply that the cited meaning of the term "prop" is a candidate for confusion in the context of a test case when we use "proptest" all over the place and the "prop" in that term means something else.

Reply:
Fair point. Also arguing for the sake of arguing now, that context of using “proptest” to mean “test the property” all over the place does not improve the quality of name “prop_add”. Thanks for changing it. :+1: